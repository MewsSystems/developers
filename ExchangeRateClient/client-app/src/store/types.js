// @flow strict

export type Data = {|
  [id: string]: {|
    currencies: Array<{| +code: string, +name: string |}>,
  |},
|};

export type SelectedRates = string[];

export type Rates = {| [id: string]: {| +current: number, +before: number |} |};

export type StateTypes = {
  +isFetchingConfig: boolean,
  +isFetchingRates: boolean,
  +rates: Rates,
  +selectedRates: SelectedRates,
  +data: Data,
  +APIerror: ?Error,
};
