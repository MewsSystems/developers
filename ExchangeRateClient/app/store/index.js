export { store, persistor } from './configureStore';
