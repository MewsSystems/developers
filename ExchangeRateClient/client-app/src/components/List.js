// @flow strict

import * as React from "react";
import styled from "styled-components";
import { connect } from "react-redux";
import Rate from "./Rate";
import createRateLabel from "../utils/createRateLabel";
import defaultTheme from "../theme";
import type { Theme } from "../theme/types";
import type { Rates, Data } from "../store/types";

const Item = styled.li`
  display: flex;
  align-items: center;
  justify-content: flex-start;
  font-size: 2rem;
  padding: 0.5rem 1rem;
  opacity: 0.7;
  background-color: ${({ isOdd, theme }: { isOdd: boolean, theme: Theme }) =>
    isOdd ? theme.colors.color6 : "grey"};
`;

Item.defaultProps = {
  theme: defaultTheme,
};

const ListContainer = styled.ul`
  margin: 2em;
  list-style-type: none;
`;

const Sublabel = styled.span`
  font-size: 15px;
  color: ${({ theme }: { theme: Theme }) => theme.colors.white};
`;

Sublabel.defaultProps = {
  theme: defaultTheme,
};

const Label = styled.span`
  flex: 1;
  display: flex;
  flex-direction: column;
`;

type Props = {
  data: Data,
  rates: Rates,
};

const List = ({ data, rates }: Props) => (
  <ListContainer>
    {Object.keys(rates).map((id, index) => (
      <Item key={id} isOdd={index % 2 === 0}>
        <Label>
          {`${createRateLabel(data[id].currencies)}:`}
          <Sublabel>
            {data[id].currencies[0].name} - {data[id].currencies[1].name}
          </Sublabel>
        </Label>
        <Rate current={rates[id].current} before={rates[id].before} />
      </Item>
    ))}
  </ListContainer>
);

const mapStateToProps = ({ data, rates }) => ({
  rates,
  data,
});

export default connect(mapStateToProps)(List);
