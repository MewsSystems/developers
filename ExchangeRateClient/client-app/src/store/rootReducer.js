// @flow

import {
  CONFIG_FETCH_SUCCESS,
  CONFIGS_FETCH_START,
  CONFIGS_FETCH_FAIL,
  RATES_FETCH_START,
  RATES_FETCH_SUCCESS,
  RATES_FETCH_FAIL,
  SELECT_RATES_IDS,
  type Action,
} from "../actions/configActions";
import type { StateTypes } from "./types";

export const initalState: StateTypes = {
  selectedRates: [],
  data: {},
  isFetchingConfig: false,
  APIerror: null,
  APIerror: null,
  isFetchingRates: false,
  rates: {},
};

export default (state: StateTypes = { ...initalState }, action: Action) => {
  switch (action.type) {
    case CONFIG_FETCH_SUCCESS:
      return {
        ...state,
        isFetchingConfig: false,
        ...action.payload,
      };
    case CONFIGS_FETCH_START:
      return {
        ...state,
        isFetchingConfig: true,
      };
    case CONFIGS_FETCH_FAIL:
      return {
        ...initalState,
        APIerror: action.payload.error,
      };
    case RATES_FETCH_START:
      return {
        ...state,
        selectedRates: action.payload,
        isFetchingRates: true,
      };
    case RATES_FETCH_SUCCESS: {
      const rates = action.payload;
      const ids = Object.keys(rates);

      const temp = ids.reduce(
        (acc, id) => ({
          ...acc,
          [id]: {
            current: rates[id],
            before: state.rates && state.rates[id] && state.rates[id].current,
          },
        }),
        {},
      );

      return {
        ...state,
        rates: temp,
        isFetchingRates: false,
      };
    }
    case RATES_FETCH_FAIL:
      return {
        ...state,
        isFetchingRates: false,
        APIerror: action.payload.error,
      };

    case SELECT_RATES_IDS:
      return {
        ...state,
        selectedRates: action.payload.ids,
      };

    default:
      return state;
  }
};
