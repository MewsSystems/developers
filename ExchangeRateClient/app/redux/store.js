import { createStore } from 'redux';
import rootReducer from './reducers';

// set initial state
const initialState = {
  rates: {},
  pairs: {},
  trends: {},
};

export default createStore(
  rootReducer,
  initialState /* eslint-disable-next-line */,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
);
