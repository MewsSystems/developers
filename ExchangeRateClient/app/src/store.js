import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './reducers';

const initialState = {};

const middleware = [thunk];

const store = createStore(
  rootReducer,
  initialState,
  compose(
    applyMiddleware(...middleware),
  )
);

export default store;
