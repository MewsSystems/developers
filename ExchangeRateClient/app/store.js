import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import logger from "redux-logger";
import rootReducer from "./reducers";

const initialState = {};

const middleware = [thunk, logger];

const store = createStore(
  rootReducer,
  initialState,
  compose(
    applyMiddleware(...middleware),
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  )
);

// if (module.hot) {
//   module.hot.accept("./reducers", () => {
//     const nextRootReducer = require("./reducers/index").default;
//     store.replaceReducer(nextRootReducer);
//   });
// }

export default store;
