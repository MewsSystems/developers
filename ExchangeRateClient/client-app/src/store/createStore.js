// @flow strict

import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import rootReducer from "./rootReducer";

export default function configureStore() {
  return createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)));
}
