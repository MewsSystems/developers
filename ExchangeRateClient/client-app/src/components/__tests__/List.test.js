// @flow strict
import { cleanup } from "react-testing-library";
import React from "react";
import List from "../List";
import renderWithRedux from "../../utils/testsUtils";
import store from "../../records/storeState";

afterEach(cleanup);

jest.mock("../Rate", () => ({ before, current }: { before: ?number, current: number }) => (
  <span>
    {before} {current}
  </span>
));

describe("List", () => {
  test("render", () => {
    const { baseElement } = renderWithRedux(<List />, { initialState: store });
    expect(baseElement).toMatchSnapshot();
  });
});
