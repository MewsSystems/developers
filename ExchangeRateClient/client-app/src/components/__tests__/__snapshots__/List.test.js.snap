// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List render 1`] = `
<body>
  <div>
    <ul
      class="sc-bwzfXH jWdLoo"
    >
      <li
        class="sc-bdVaJa hDbPb"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          CZK - LSL:
          <span
            class="sc-htpNat jAkoja"
          >
            Czech Republic Koruna
            - 
            Lesotho Loti
          </span>
        </span>
        <span>
          2.6048
          2.6136
        </span>
      </li>
      <li
        class="sc-bdVaJa ihlOSe"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          PHP - DZD:
          <span
            class="sc-htpNat jAkoja"
          >
            Philippines Peso
            - 
            Algeria Dinar
          </span>
        </span>
        <span>
          2.1716
          2.1669
        </span>
      </li>
      <li
        class="sc-bdVaJa hDbPb"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          LSL - AUD:
          <span
            class="sc-htpNat jAkoja"
          >
            Lesotho Loti
            - 
            Australia Dollar
          </span>
        </span>
        <span>
          2.3991
          2.4023
        </span>
      </li>
      <li
        class="sc-bdVaJa ihlOSe"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          QAR - JOD:
          <span
            class="sc-htpNat jAkoja"
          >
            Qatar Riyal
            - 
            Jordan Dinar
          </span>
        </span>
        <span>
          2.8342
          2.8389
        </span>
      </li>
      <li
        class="sc-bdVaJa hDbPb"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          HKD - FKP:
          <span
            class="sc-htpNat jAkoja"
          >
            Hong Kong Dollar
            - 
            Falkland Islands (Malvinas) Pound
          </span>
        </span>
        <span>
          2.8459
          2.8539
        </span>
      </li>
      <li
        class="sc-bdVaJa ihlOSe"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          DZD - MVR:
          <span
            class="sc-htpNat jAkoja"
          >
            Algeria Dinar
            - 
            Maldives (Maldive Islands) Rufiyaa
          </span>
        </span>
        <span>
          1.7488
          1.756
        </span>
      </li>
      <li
        class="sc-bdVaJa hDbPb"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          AMD - GEL:
          <span
            class="sc-htpNat jAkoja"
          >
            Armenia Dram
            - 
            Georgia Lari
          </span>
        </span>
        <span>
          0.8003
          0.8069
        </span>
      </li>
      <li
        class="sc-bdVaJa ihlOSe"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          SRD - CZK:
          <span
            class="sc-htpNat jAkoja"
          >
            Suriname Dollar
            - 
            Czech Republic Koruna
          </span>
        </span>
        <span>
          2.3798
          2.3794
        </span>
      </li>
      <li
        class="sc-bdVaJa hDbPb"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          BIF - CLP:
          <span
            class="sc-htpNat jAkoja"
          >
            Burundi Franc
            - 
            Chile Peso
          </span>
        </span>
        <span>
          2.4028
          2.412
        </span>
      </li>
      <li
        class="sc-bdVaJa ihlOSe"
      >
        <span
          class="sc-bxivhb iOtzTq"
        >
          BDT - VEF:
          <span
            class="sc-htpNat jAkoja"
          >
            Bangladesh Taka
            - 
            Venezuela Bolivar
          </span>
        </span>
        <span>
          2.7308
          2.7211
        </span>
      </li>
    </ul>
  </div>
</body>
`;
